xdsl/interpreters/eqsat_pdl_interp.py

```diff
@@ -270,25 +265,46 @@ def run_replace(
                 "Replacement value must be the result of an EClassOp"
             )
 
-        repl_eclass = self.eclass_union_find.find(repl_eclass)
-        original_eclass = self.eclass_union_find.find(original_eclass)
+        if self.eclass_union(original_eclass, repl_eclass):
+            self.worklist.append(original_eclass)
 
-        if repl_eclass == original_eclass:
-            return ()
+        return ()
+
+    def eclass_union(self, a: eqsat.EClassOp, b: eqsat.EClassOp) -> bool:
+        """Unions two e-classes, merging their operands and results.
+        Returns True if the e-classes were merged, False if they were already the same."""
+        a = self.eclass_union_find.find(a)
+        b = self.eclass_union_find.find(b)
+
+        if a == b:
+            return False
 
         self.eclass_union_find.union(
-            original_eclass,
-            repl_eclass,
+            a,
+            b,
         )
-        if self.eclass_union_find.find(original_eclass) == repl_eclass:
-            # In the union-find the canonical representative of the original_eclass
-            # is now the repl_eclass, so we have to keep the repl_eclass:
-            self.merge_list.append(MergeTodo(repl_eclass, original_eclass))
+
+        if self.eclass_union_find.find(a) == a:
+            to_keep = a
+            to_replace = b
         else:
-            # otherwise we keep the original_eclass:
-            self.merge_list.append(MergeTodo(original_eclass, repl_eclass))
+            assert self.eclass_union_find.find(b) == b
+            to_keep = b
+            to_replace = a
 
-        return ()
+        # Operands need to be deduplicated because it can happen the same operand was
+        # used by different parent eclasses after their children were merged:
+        new_operands = OrderedSet((*to_keep.operands, *to_replace.operands))
```

Comment:
Seems better to avoid creating a temporary tuple
```suggestion
        new_operands = OrderedSet()
        new_operands.update(to_keep.operands)
        new_operands.update(to_replace.operands)
```